[Huggenguggen] iP

Huggenguggen opened the pull request:
# DukePro
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))
DukePro frees your mind of having to remember things you need to do. It's, 
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do it,
1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html). 
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [ ] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Huggenguggen committed:
Level-1 v1

Added greet, echo and end_Program methods for a basic skeleton

Huggenguggen committed:
Level 2 Done

added add to list and print list functionality

Huggenguggen committed:
Level 3 done

Turned items into class Task
Added ability to mark items as done.

Huggenguggen committed:
Level-4 done

Added support for 3 types of tasks:
1. ToDos
2. Deadlines
3. Events

Note:
There is still support for original tasks (i.e. any text that are not keywords)

Huggenguggen committed:
A-TextUITesting

Changed some formatting for easier diffing
Modified text-ui-test files for semi-automated testing

Huggenguggen committed:
Level-5

Added error handling via DukeException
Some more specific exceptions as well such as AlreadyMarkedException and AlreadyUnmarkedException

Huggenguggen committed:
Level-6

Added delete functionality
Refactored parse_cmd

Huggenguggen committed:
Level-7 done

Added persistent storage in hard disk for tasks. AI-Chan will load the data from the hard disk when started up. Data can be found in [project_root]/data/duke.txt

Huggenguggen committed:
Level-8 done

Redid parsing of commands to account for dates. Added understanding of dates and times for events and deadlines. As a consequence, users must now input in a certain format.

Huggenguggen committed:
Merge branch 'branch-Level-8'

Huggenguggen committed:
Bug fixes

Fixed bug which emerged due to merging of Level-7 and Level-8 branches. The persistent data was being saved but when loaded was coming up as null. It is now fixed and LocalDateTime should work on the persistent data as well.

Huggenguggen committed:
A-MoreOOP

Extracted methods out to classes, Ui, Storage, Parser, TaskList

Huggenguggen committed:
A-Packages

divided classes into packages under the duke package

Huggenguggen committed:
Merge branch 'add-gradle-support'

Huggenguggen committed:
A-Gradle and A-JUnit

Added gradle support along with junit support and some simple tests for ToDo and TaskList

Huggenguggen committed:
Create MANIFEST.MF

Huggenguggen committed:
Add JavaDoc to most methods

Huggenguggen committed:
Tweak code to comply with coding standard

Huggenguggen committed:
Implement Level 9

Huggenguggen committed:
Merge branch 'branch-A-CodingStandard'

Huggenguggen committed:
Merge branch 'branch-Level-9'

Huggenguggen committed:
Bug Fix

Huggenguggen committed:
Add Checkstyle support

Huggenguggen committed:
Implement Level-10

Huggenguggen committed:
Merge branch 'branch-Level-10'

gremmyz commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,105 @@
+            curr = new String[]{curr_in};
+        }
+        String[] curr_title = curr[0].split(" "); //split title by word
+        if (curr_in.equals("bye")) {
+            taskList.end(ui);
+        } else if (curr_in.equals("list")) {
+            ui.print(taskList);
+        } else if (curr_title[0].equals("mark")) {
+            taskList.mark(curr_title, ui);
+        } else if (curr_title[0].equals("unmark")) {
+            taskList.unmark(curr_title, ui);
+        } else if (curr_title[0].equals("todo")) {
+            taskList.addToDo(curr, ui);
+        } else if (curr_title[0].equals("deadline")) {
+            taskList.addDeadline(curr, ui);
+        } else if (curr_title[0].equals("event")) {
+            taskList.addEvent(curr, ui);
+        } else if (curr_title[0].equals("delete")) {
+            taskList.deleteTask(curr_title, ui);
+        } else if (curr_title[0].equals("find")) {
+            taskList.findTask(curr_title, ui);
+        } else {
+            throw new DukeException("Hmmm, I don't understand what you want to do");
```
Any reason why you chose to use _if-else_ statements instead of _switch_?

Huggenguggen replied:
At the time I felt it was quite an unnecessary abstraction to make an enumeration and hence _switch_ statement. I might consider it in the future.

gremmyz commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,74 @@
+
+public class Task {
+    boolean done;
+    private final String CHECKED = "[X]";
+    private final String UNCHECKED = "[ ]";
```
Good job with the constants naming! 👍🏼

gremmyz commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,403 @@
+package duke.task;
+
+import duke.exception.*;
+import duke.ui.*;
```
As mentioned above, I think the imported classes should be listed explicitly as per the coding standard given?
```suggestion
import duke.exception.DukeException;
import duke.ui.Ui;
```

gremmyz commented on src/main/java/duke/task/ToDo.java:
```diff
@@ -0,0 +1,37 @@
+        this.type = "[T]";
+    }
+
+    /**
+     * Overloaded Constructor for when loading from hard drive.
+     *
+     * @param title title of the ToDo.
+     * @param done boolean to represent if task is done or not;
+     */
+    public ToDo(String title, boolean done) {
+        super(title);
+        this.type = "[T]";
+        this.done = done;
+    }
```
I'm not sure what this does, may I know how it works?

Huggenguggen replied:
Yep sure! This is an overloaded constructor for when I read the data from the stored data. I need it for setting the status of the task on initialization since my original constructor would default to not done.

gremmyz commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,102 @@
+    private TaskList taskList;
+    private Storage storage;
+    private Ui ui;
+    private boolean hasGui;
```
This is cool, users have the option to choose to use either the CLI or GUI?

Huggenguggen replied:
This is a future feature mostly aimed for developer users/myself to toggle between CLI and GUI. I'm not entirely sure if it's feasible to let users have the option to switch from CLI to GUI and vice versa on the fly.

gremmyz commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,71 @@
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.Scanner;
+import duke.task.*;
```
I think the imported classes should be listed explicitly as per the coding standard given?
```suggestion
import duke.task.Deadline;
import duke.task.Event;
import duke.task.Task;
import duke.task.TaskList;
import duke.task.ToDo;
```

gremmyz reviewed (commented):
Looks good! 👍🏼 Really clean code overall, well done. I like the concept of it too :D

Huggenguggen committed:
Update src/main/java/duke/task/TaskList.java

Co-authored-by: Ng Jia Yi <92777033+gremmyz@users.noreply.github.com>

Huggenguggen committed:
Update src/main/java/duke/storage/Storage.java

Co-authored-by: Ng Jia Yi <92777033+gremmyz@users.noreply.github.com>

Huggenguggen committed:
Change styling of most files

Massive styling overhaul

As pointed out in code review my styling could be more consistent.

Have changed most files to comply with the checkstyle and hence styling standards.

Huggenguggen committed:
Add assertions

Added assertions to enforce behavior

Used assert to document important assumptions that should hold at some points.

Huggenguggen committed:
refactor: Collapse if else

Collapse the unnecessary if else into simple if statements.

These are logically equivalent and makes it easier to see the good path.

Huggenguggen committed:
Merge pull request #1 from Huggenguggen/branch-A-Assertions

Add assertions

Huggenguggen committed:
Merge pull request #2 from Huggenguggen/branch-A-CodeQuality

refactor: Collapse if else

Huggenguggen committed:
Update Task.java

Update Task.java to support better searching

niekis commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,71 @@
+ */
+public class Deadline extends Task {
+    private String deadline;
+    private LocalDateTime dateline;
```
Does `dateline` refer to the DateTime representation of the task's deadline? Personally, I think that this might be confusing as dateline has another meaning all together. I would suggest changing the names of the two attributes from deadline -> deadlineString and dateline -> deadlineDatetime!

niekis commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,213 @@
+/**
+ * Class for handling UI (user facing interface)
+ */
+public class Ui {
```
Cool logo!

niekis commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,213 @@
+    /**
+     * Method for first greet message.
+     */
+    public void greet() {
```
Since you have two sets function for gui and non-gui, would it be clearer if you specificed the context of the non-gui functions in their names? (ie. defaultGreet, defaultDeleteMessage, etc)

niekis commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,213 @@
+     * @param temp taskList to be printed.
+     */
+    public void printSearch(TaskList temp) {
+        System.out.println(LINE_BREAK + "\tHere are the matching tasks!");
```
Question: is there a specific reason why you did not combine these 3 lines into one println statement?

niekis commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,105 @@
+     * @param taskList TaskList containing list of all tasks in the program.
+     * @param ui ui to show user messages.
+     * @throws DukeException if there is an error (usually user input error).
+     */
```
Wow, I really like how this function is organised! Great job being careful with nested code!

niekis reviewed (commented):
LGTM!! I can only find a few nits. Good job!

Huggenguggen committed:
Tweak Styling

Made some changes to the styling of Text fields and the overall GUI to be more pleasing to the eye.

Huggenguggen committed:
Add User Guide

Huggenguggen committed:
Create _config.yml

Huggenguggen committed:
Update README.md

Remove unnecessary line breaks

Huggenguggen committed:
Update README.md

Add extra detail for Expected Outcome of exit command.

Add more detail to acknowledgements segment.
